tmt/utils.py

```diff
@@ -410,6 +410,249 @@ class CommandOutput(NamedTuple):
+        except subprocess.CalledProcessError:
+            # Interactive mode can return non-zero if the last command
+            # failed, ignore errors here
+            pass
+        finally:
+            return CommandOutput(None, None)
+
+    # Create the process
+    try:
+        process = subprocess.Popen(
+            command.to_popen(),
+            cwd=cwd, shell=shell, env=environment,
+            start_new_session=True,
+            stdin=subprocess.DEVNULL, stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT if join else subprocess.PIPE,
+            executable=executable)
+    except FileNotFoundError as error:
+        raise RunError(
+            f"File '{error.filename}' not found.",
+            command,
+            127,
+            caller=caller)
+
+    stdout_thread = StreamLogger(
+        process.stdout,
+        log_header='out',
+        logger=log,
+        click_context=click.get_current_context(silent=True))
+    stderr_thread = stdout_thread
+    if not join:
+        stderr_thread = StreamLogger(
+            process.stderr,
+            log_header='err',
+            logger=log,
+            click_context=click.get_current_context(silent=True))
+    stdout_thread.start()
+    if not join:
+        stderr_thread.start()
+
+    # A bit of logging helpers for debugging duration behavior
+    start_timestamp = time.monotonic()
+
+    def _event_timestamp() -> str:
+        return f'{time.monotonic() - start_timestamp:.4}'
+
+    def log_event(msg: str) -> None:
+        logger.debug('Command event', f'{_event_timestamp()} {msg}', level=4)
+
+    try:
+        process.wait(timeout=timeout)
+    except subprocess.TimeoutExpired:
+        log_event(f'duration "{timeout}" exceeded')
+
+        os.killpg(os.getpgid(process.pid), signal.SIGKILL)
+        log_event('sent SIGKILL signal')
+
+        process.wait()
+        log_event('kill confirmed')
+
+        process.returncode = PROCESS_TIMEOUT
+
+    log_event('waiting for stream readers')
+
+    stdout_thread.join()
+    log_event('stdout reader done')
+
+    if not join:
+        stderr_thread.join()
+        log_event('stderr reader done')
+
+    # Handle the exit code, return output
+    if process.returncode != 0:
+        raise RunError(
+            message=f"Command returned '{process.returncode}'.",
+            command=command,
+            returncode=process.returncode,
+            stdout=stdout_thread.get_output(),
+            stderr=stderr_thread.get_output(),
+            caller=caller)
+    if join:
+        return CommandOutput(
+            stdout_thread.get_output(), None)
+    else:
+        return CommandOutput(
+            stdout_thread.get_output(), stderr_thread.get_output())
+
+
+def run_command(
+        *,
+        command: Command,
+        friendly_command: Optional[str] = None,
+        message: Optional[str] = None,
+        silent: bool = False,
+        cwd: Optional[Path] = None,
+        dryrun: bool = False,
```

Comment:
Can we keep the parameter name as `dry`, consistent with the `--dry` option name? I know, the original meaning in `Common.run()` was different, but here it is quite straightforward, I'd say.

Reply:
Fixed in fb3d34af.